src/fighter_state.rs

```diff
@@ -0,0 +1,747 @@
+impl Grabbing {
+    pub const PRIORITY: i32 = Throwing::PRIORITY;
+}
+
+/// Component indicating the player is flopping
+#[derive(Component, Reflect, Default, Debug)]
+#[component(storage = "SparseSet")]
+pub struct Attacking {
+    /// The initial y-height of the figther when starting the attack
+    pub start_y: f32,
+    pub has_started: bool,
+    pub is_finished: bool,
+}
+impl Attacking {
+    pub const PRIORITY: i32 = 30;
+    pub const ANIMATION: &'static str = "attacking";
+}
+
+/// Component indicating the player is getting knocked back
+#[derive(Component, Reflect, Default, Debug)]
+#[component(storage = "SparseSet")]
+pub struct KnockedBack {
+    pub velocity: Vec2,
+    pub timer: Timer,
+}
+impl KnockedBack {
+    pub const PRIORITY: i32 = 20;
+    pub const ANIMATION_LEFT: &'static str = "knocked_left";
+    pub const ANIMATION_RIGHT: &'static str = "knocked_right";
+}
+
+/// Component indicating the player is dying
+#[derive(Component, Reflect, Default, Debug)]
+#[component(storage = "SparseSet")]
+pub struct Dying;
+impl Dying {
+    pub const PRIORITY: i32 = 1000;
+    pub const ANIMATION: &'static str = "dying";
+}
+
+//
+// Fighter input collector systems
+//
+
+/// Emits state transitions based on fighter actions
+fn collect_player_actions(
+    mut players: Query<
+        (
+            &ActionState<PlayerAction>,
+            &mut StateTransitionIntents,
+            &Inventory,
+            &Stats,
+        ),
+        With<Player>,
+    >,
+) {
+    for (action_state, mut transition_intents, inventory, stats) in &mut players {
+        // Trigger attacks
+        if action_state.just_pressed(PlayerAction::Attack) {
+            transition_intents.push_back(StateTransition::new(
+                Attacking::default(),
+                Attacking::PRIORITY,
+                false,
+            ));
+        }
+
+        // Trigger grab/throw
+        if action_state.just_pressed(PlayerAction::Throw) {
+            if inventory.is_some() {
+                transition_intents.push_back(StateTransition::new(
+                    Throwing,
+                    Throwing::PRIORITY,
+                    true,
+                ));
+            } else {
+                transition_intents.push_back(StateTransition::new(
+                    Grabbing,
+                    Grabbing::PRIORITY,
+                    true,
+                ));
+            }
+        }
+
+        // Trigger movement
+        if action_state.pressed(PlayerAction::Move) {
+            let dual_axis = action_state.clamped_axis_pair(PlayerAction::Move).unwrap();
+            let direction = dual_axis.xy();
+
+            transition_intents.push_back(StateTransition::new(
+                Moving {
+                    velocity: direction * stats.movement_speed,
+                },
+                Moving::PRIORITY,
+                false,
+            ));
+        }
+    }
+}
+
+/// Look for attacks that have contacted a figher and knock them back
```

Comment:
Good in this PR but leaving a note that future design intent will require not all attacks to apply knockback.

Reply:
:+1:

Not for this PR, but here's a thought:

I already have a `damage_velocity` field on the `Damage` component.

Maybe instead we should change `damage_velocity` to `damage_impuse` and make it a new struct like this:

```rust
struct Impulse {
    velocity: Vec2,
    time: f32,
}
```

Then if you made an attack where the damage_impulse was `Impulse::default()` which has a `time` and `velocity` of 0.0, this system would check that, and in that case it wouldn't put the player into a knockback state.